File: um_microservice/src/types/secrets_initializer.go
package types

import (
	"log"
	"os"
)

type SecretInitializer struct{}

var instance *SecretInitializer

func NewSecretInitializer() *SecretInitializer {
	if instance == nil {
		instance = &SecretInitializer{}
	}
	return instance
}

func (si *SecretInitializer) InitSecrets() {
	si.initSecret("PASSWORD")
}

func (si *SecretInitializer) initSecret(envVarName string) {
	secretPath := os.Getenv(envVarName)
	secretValue, err := os.ReadFile(secretPath)
	if err != nil {
		log.Fatalf("Error reading secret file for %s: %v", envVarName, err)
	}
	err = os.Setenv(envVarName, string(secretValue))
	if err != nil {
		log.Printf("Initialized %s.\n", envVarName)
		return
	}
}

File: worker_microservice/worker/classes/secrets_initializer.py
import os
from worker.utils.logger import logger


# implementing pattern Singleton for SecretInitializer class
class SecretInitializer:

    _instance = None  # starter value: no object initially instantiated

    # if SecretInitializer object already instantiated, then return it without re-instantiating
    def __new__(cls):
        if not cls._instance:
            cls._instance = super().__new__(cls)
        return cls._instance

    # setting env variables for secrets
    def init_secrets(self):
        self.__init_secret('PASSWORD')
        self.__init_secret('APIKEY')

    @staticmethod
    def __init_secret(env_var_name):
        secret_path = os.environ.get(env_var_name)
        with open(secret_path, 'r') as file:
            secret_value = file.read()
        os.environ[env_var_name] = secret_value
        logger.info(f"Initialized {env_var_name}.\n")